contracts/cw1-whitelist-ng/src/state.rs

```diff
@@ -0,0 +1,91 @@
+use std::marker::PhantomData;
+
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use cosmwasm_std::{Addr, Empty};
+use cw_storage_plus::Item;
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug, Default)]
+pub struct AdminList {
+    pub admins: Vec<Addr>,
+    pub mutable: bool,
+}
+
+impl AdminList {
+    /// returns true if the address is a registered admin
+    pub fn is_admin(&self, addr: &str) -> bool {
+        self.admins.iter().any(|a| a.as_ref() == addr)
+    }
+
+    /// returns true if the address is a registered admin and the config is mutable
+    pub fn can_modify(&self, addr: &str) -> bool {
+        self.mutable && self.is_admin(addr)
+    }
+}
+
+pub struct Cw1WhitelistContract<T> {
+    // I am pretty sure that just form this with some proper hint attributes it would be possible
+    // to provide helpers for raw queries, this might be fun idea
+    pub(crate) admin_list: Item<'static, AdminList>,
+    _msg: PhantomData<T>,
+}
+
+impl Cw1WhitelistContract<Empty> {
```

Comment:
Okay, this is to fix the T we use in this contract as opposed to the generic T we allow for anyone who imports as a library to extend.

I see the point. I wonder if this makes sense to be in lib.rs next to the entry points.
Some contract importing it would just need to copy lib.rs and edit it to allow custom messages.